cms/core/blocks/embeddable.py

```diff
@@ -98,3 +100,86 @@ def clean(self, value: "StructValue") -> "StructValue":
     class Meta:
         icon = "code"
         template = "templates/components/streamfield/ons_embed_block.html"
+
+
+class VideoEmbedBlock(blocks.StructBlock):
+    """A video embed block."""
+
+    link_url = blocks.URLBlock(
+        help_text=_(
+            "The URL to the video hosted on YouTube or Vimeo, for example, "
+            "https://www.youtube.com/watch?v={ video ID } or https://vimeo.com/video/{ video ID }. "
+            "Used to link to the video when cookies are not enabled."
+        )
+    )
+    image = ImageChooserBlock(help_text=_("The video cover image, used when cookies are not enabled."))
+    title = blocks.CharBlock(help_text=_("The descriptive title for the video used by screen readers."))
+    link_text = blocks.CharBlock(
+        help_text=_("The text to be shown when cookies are not enabled e.g. 'Watch the {title} on Youtube'.")
+    )
+
+    def get_context(self, value: "StreamValue", parent_context: dict | None = None) -> dict:
+        """Get the embed URL for the video based on the link URL."""
+        context: dict = super().get_context(value, parent_context=parent_context)
+        embed_url = ""
+        # Vimeo
+        if urlparse(value["link_url"]).hostname in ["www.vimeo.com", "vimeo.com", "player.vimeo.com"]:
+            url_path = urlparse(value["link_url"]).path.strip("/")
+            # Handle different Vimeo URL patterns
+            if "video/" in url_path:  # noqa: SIM108
+                # Handle https://vimeo.com/showcase/7934865/video/ID format
+                video_id = url_path.split("video/")[-1]
+            else:
+                # Handle https://player.vimeo.com/video/ID or https://vimeo.com/ID format
+                video_id = url_path.split("/")[0]
+            # Remove any query parameters from video ID
+            video_id = video_id.split("?")[0]
+            embed_url = "https://player.vimeo.com/video/" + video_id
+        # YouTube
+        elif urlparse(value["link_url"]).hostname in ["www.youtube.com", "youtube.com", "youtu.be"]:
+            url_parts = urlparse(value["link_url"])
+            if url_parts.hostname == "youtu.be":
+                # Handle https://youtu.be/ID format
+                video_id = url_parts.path.lstrip("/")
+            elif "/v/" in url_parts.path:
+                # Handle https://www.youtube.com/v/ID format
+                video_id = url_parts.path.split("/v/")[-1]
+            else:
+                # Handle https://www.youtube.com/watch?v=ID format
+                query = dict(param.split("=") for param in url_parts.query.split("&"))
+                video_id = query.get("v", "")
+            # Remove any query parameters from video ID
+            video_id = video_id.split("?")[0]
+            embed_url = "https://www.youtube.com/embed/" + video_id
+        context["value"]["embed_url"] = embed_url
```

Comment:
suggestion: extract the embed_url extraction to its own method so that `get_context` becomes:

```python
def get_context(self, value, parent_context=None):
    context = super().get_context(value, parent_contex=parent_context)

    context["embed_url"] = self.get_embed_url(value["link_url"])
```

also, with this you can just use `embed_url` in the template

Reply:
Done.